h5py/_hl/filters.py

```diff
@@ -345,34 +345,39 @@ def guess_chunk(shape, maxshape, typesize):
     """
     # pylint: disable=unused-argument
 
+    is_unlimited = np.array([x == 0 for x in shape])
+
     # For unlimited dimensions we have to guess 1024
-    shape = tuple((x if x!=0 else 1024) for i, x in enumerate(shape))
+    chunks = np.array([x if x!=0 else 1024 for x in shape], dtype='=f8')
 
     ndims = len(shape)
     if ndims == 0:
         raise ValueError("Chunks not allowed for scalar datasets.")
 
-    chunks = np.array(shape, dtype='=f8')
     if not np.all(np.isfinite(chunks)):
         raise ValueError("Illegal value in chunk tuple")
 
     # Determine the optimal chunk size in bytes using a PyTables expression.
     # This is kept as a float.
-    dset_size = np.product(chunks)*typesize
+    dset_size = np.product(chunks[~is_unlimited])*typesize
```

Comment:
What happens if you do not exclude the unlimited dimensions from this target?  This is currently producing much smaller chunks than master branch does.

Reply:
I thought it would be nice for the chunks to match in this case:

```python
>>> guess_chunk((0, 250, 250, 120), None, 8)
(1, 16, 32, 15)  # master gives (32, 16, 16, 8)
>>> guess_chunk((1, 250, 250, 120), None, 8)
(1, 16, 32, 15)  # master gives (1, 16, 32, 15)
```

So this way for large fixed dimensions, adding an unlimited dimension is the same as adding a fixed dimension of size 1 (in terms of the resulting chunks).

Without this, you get that:

```python
>>> guess_chunk((0, 250, 250, 120), None, 8)
(1, 32, 63, 30)
>>> guess_chunk((1, 250, 250, 120), None, 8)
(1, 16, 32, 15)
>>> guess_chunk((0,), None, 8)
(1024,)
>>> guess_chunk((0,) * 10, None, 8)
(2, 2, 2, 2, 4, 4, 4, 4, 4, 4)
>>> guess_chunk((0, 1), None, 8)
(1024, 1)
>>> guess_chunk((2,) * 20 + (0,), None, 8)
(1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1)
```

Not bad either, so if you prefer that we can change it.